src/_P164_gases_ens160.ino

```diff
@@ -0,0 +1,158 @@
+      strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[1], PSTR(PLUGIN_VALUENAME2_164));
+      break;
+    }
+
+    case PLUGIN_I2C_HAS_ADDRESS:
+    case PLUGIN_WEBFORM_SHOW_I2C_PARAMS:
+    {
+      const uint8_t i2cAddressValues[] = { ENS160_I2CADDR_0, ENS160_I2CADDR_1 };
+      constexpr int nrAddressOptions   = NR_ELEMENTS(i2cAddressValues);
+
+      if (function == PLUGIN_WEBFORM_SHOW_I2C_PARAMS) {
+        addFormSelectorI2C(F("i2c_addr"), nrAddressOptions, i2cAddressValues, P164_I2C_ADDR);
+        addFormNote(F("ADDR Low=0x52, High=0x53"));
+      } else {
+        success = intArrayContains(nrAddressOptions, i2cAddressValues, event->Par1);
+      }
+
+      break;
+    }
+
+    # if FEATURE_I2C_GET_ADDRESS
+    case PLUGIN_I2C_GET_ADDRESS:
+    {
+      event->Par1 = P164_I2C_ADDR;
+      success     = true;
+      break;
+    }
+    # endif // if FEATURE_I2C_GET_ADDRESS
+
+    case PLUGIN_SET_DEFAULTS:
+    {
+      P164_I2C_ADDR = ENS160_I2CADDR_1;
+      success = true;
+      break;
+    }
+
+    case PLUGIN_INIT:
+    {
+      initPluginTaskData(event->TaskIndex, new (std::nothrow) P164_data_struct(event));
+      P164_data_struct *P164_data = static_cast<P164_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      success = (nullptr != P164_data && P164_data->begin());
+      break;
+    }
+
+    case PLUGIN_READ:
+    {
+      P164_data_struct *P164_data = static_cast<P164_data_struct *>(getPluginTaskData(event->TaskIndex));
+
+      if (nullptr == P164_data) {
+        Serial.print("P164: plugin_read NULLPTR");
+        break;
+      }
+
+      float temperature = 20.0f;  // A reasonable value in case temperature source task is invalid
+      float humidity = 50.0f;     // A reasonable value in case humidity source task is invalid
+      if (validTaskIndex(P164_PCONFIG_TEMP_TASK) && validTaskIndex(P164_PCONFIG_HUM_TASK))
+      {
+        // we're checking a var from another task, so calculate that basevar
+        temperature = UserVar[P164_PCONFIG_TEMP_TASK * VARS_PER_TASK + P164_PCONFIG_TEMP_VAL]; // in degrees C
+        humidity = UserVar[P164_PCONFIG_HUM_TASK * VARS_PER_TASK + P164_PCONFIG_HUM_VAL];    // in % relative
+      }
+      success = P164_data->read(UserVar[event->BaseVarIndex], UserVar[event->BaseVarIndex + 1], temperature, humidity);
+
+      success = true;
+      break;
+    }
+
+    case PLUGIN_WEBFORM_LOAD:
+    {
+      success = P164_data_struct::webformLoad(event);
+      break;
+    }
+
+    case PLUGIN_WEBFORM_SAVE:
+    {
+      success = P164_data_struct::webformSave(event);
+      break;
+    }
+
+    case PLUGIN_TEN_PER_SECOND:
+    {
+      P164_data_struct *P164_data =
+        static_cast<P164_data_struct *>(getPluginTaskData(event->TaskIndex));
+      if (nullptr == P164_data) {
+        break;
+      }
+      success = P164_data->tenPerSecond(event);
+      success = true;
+    }
+
+    case PLUGIN_FIFTY_PER_SECOND:
+    case PLUGIN_ONCE_A_SECOND:
+    {
+      break;
+    }
```

Comment:
Unused function cases can be removed, unless you intend to implement these later? (Compiler should optimize this away, I know, but better make it explicit)

Reply:
Indeed a left over from debugging phase. I am still looking for a "life of a plugin" documentation showing when the plugin is created/destroyed and when each of the events occur. Fixed.

Reply:
The start/end of a plugin (or to be more exact the task) are: `PLUGIN_INIT` and `PLUGIN_EXIT`.

When saving settings, `PLUGIN_EXIT` is called and if the task is enabled, `PLUGIN_INIT` is called.
There is some flag in the `DeviceStruct` to explicitly define whether save should happen before exit, but you can leave this at its default as it hardly ever needs to be set to 'the other' state.